net/ip.c

```diff
@@ -0,0 +1,448 @@
+	M0_LOG(M0_DEBUG, "ip: %s to hostname: %s\n",ip, he->h_name);
+	rc = snprintf(hostname, M0_NET_IP_STRLEN_MAX, "%s", he->h_name);
+
+	if (rc >= M0_NET_IP_STRLEN_MAX)
+		M0_LOG(M0_ERROR, "Hostname too long. Reqd: %d buf_len: %d", rc,
+		       M0_NET_IP_STRLEN_MAX);
+}
+
+/**
+ * This function decodes the inet format address.
+ *  The inet address format is of type
+ *    <family>:<type>:<ipaddr/hostname_FQDN>@<port>
+ *    for example: "inet:tcp:127.0.0.1@3000",
+ *                 "inet:stream:lanl.gov@23",
+ *                 "inet6:dgram:FE80::0202:B3FF:FE1E:8329@6663"
+ * Return value: 0 in case of success.
+ *               < 0 in case of error.
+ */
+static int m0_net_ip_inet_parse(const char *name, struct m0_net_ip_addr *addr)
+{
+	uint32_t     portnum;
+	int          shift = 0;
+	int          f;
+	int          s;
+	int          rc;
+	char         ip[M0_NET_IP_STRLEN_MAX] = {};
+	char         port[M0_NET_IP_PORTLEN_MAX] = {};
+	char        *at;
+	const char  *ep_name = name;
+
+	for (f = 0; f < ARRAY_SIZE(ip_family); ++f) {
+		if (ip_family[f] != NULL) {
+			shift = strlen(ip_family[f]);
+			if (strncmp(ep_name, ip_family[f], shift) == 0)
+				break;
+		}
+	}
+	if (f >= ARRAY_SIZE(ip_family) || ep_name[shift] != ':')
+		return M0_ERR(-EINVAL);
+	ep_name += shift + 1;
+	for (s = 0; s < ARRAY_SIZE(ip_protocol); ++s) {
+		if (ip_protocol[s] != NULL) {
+			shift = strlen(ip_protocol[s]);
+			if (strncmp(ep_name, ip_protocol[s], shift) == 0)
+				break;
+		}
+	}
+	if (s >= ARRAY_SIZE(ip_protocol) || ep_name[shift] != ':')
+		return M0_ERR(-EINVAL);
+	ep_name += shift + 1;
+	at = strchr(ep_name, '@');
+	if (at == NULL)
+		return M0_ERR(-EINVAL);
+	else {
+		at++;
+		if (at == NULL || at[0] < '0' || at[0] > '9')
+			return M0_ERR(-EINVAL);
+		strcpy(port, at);
+		portnum = atoi(port);
+		M0_ASSERT(portnum < M0_NET_IP_PORT_MAX);
+		addr->nia_n.nip_port = (uint16_t)portnum;
+	}
+
+	rc = m0_net_hostname_to_ip((char *)ep_name, ip, &addr->nia_n.nip_format);
+	if (rc == 0)
+		inet_pton(f == M0_NET_IP_AF_INET ? AF_INET : AF_INET6,
+			  ip, &addr->nia_n.ip_n.sn[0]);
+
+	M0_ASSERT(strlen(name) < ARRAY_SIZE(addr->nia_p));
+	strcpy(addr->nia_p, name);
+	addr->nia_n.fmt_pvt.ia.nia_family = f;
+	addr->nia_n.fmt_pvt.ia.nia_type = s;
+
+	/* Ignore the error due to gethostbyname() as it will be retried. */
+	return rc >= 0 ? M0_RC(0) : M0_ERR(rc);
+}
+
+/**
+ * This function decodes the lnet format address and extracts the ip address and
+ * port number from it.
+ * This is also used to allocate unique transfer machine identifiers for LNet
+ * network addresses with wildcard transfer machine identifier (like
+ * "192.168.96.128@tcp:12345:31:*").
+ * Return value: 0 in case of success.
+ *               < 0 in case of error.
+ */
+static int m0_net_ip_lnet_parse(const char *name, struct m0_net_ip_addr *addr)
+{
+	char        *at = NULL;
+	int          nr;
+	int          i;
+	unsigned     pid;
+	unsigned     portal;
+	unsigned     portnum;
+	unsigned     tmid;
```

Comment:
Can we use fixed-width integer type here?

Reply:
Done